lib/src/command/add.dart

```diff
@@ -36,49 +37,77 @@ class AddCommand extends PubCommand {
+  String get description => '''Add dependencies to `pubspec.yaml`.
 
-  bool get isDev => argResults['dev'];
-  bool get isDryRun => argResults['dry-run'];
-  String? get gitUrl => argResults['git-url'];
-  String? get gitPath => argResults['git-path'];
-  String? get gitRef => argResults['git-ref'];
-  String? get hostUrl => argResults['hosted-url'];
-  String? get path => argResults['path'];
-  String? get sdk => argResults['sdk'];
+Invoking `dart pub add foo bar` will add `foo` and `bar` to `pubspec.yaml`
 
-  bool get hasGitOptions => gitUrl != null || gitRef != null || gitPath != null;
-  bool get hasHostOptions => hostUrl != null;
+with a default constraint derived from latest compatible version.''';
 
-  bool get isHosted => !hasGitOptions && path == null && path == null;
+  @override
+  String get argumentsDescription => '''
+[options] [dev:]<package>[:descriptor] [[dev:]<package>[:descriptor] ...]
+
+For example:
+  * Add a hosted dependency at newest compatible stable version:
+    `$topLevelProgram pub add foo`
+  * Add a hosted dev dependency at newest compatible stable version:
+    `$topLevelProgram pub add dev:foo`
+  * Add a hosted dependency with the given constraint
+    `$topLevelProgram pub add foo:^1.2.3`
+  * Add multiple dependencies:
+    `$topLevelProgram pub add foo dev:bar`
+  * Add a path dependency:
+    `$topLevelProgram pub add 'foo{"path":"../foo"}'`
+  * Add a hosted dependency:
+    `$topLevelProgram pub add 'foo{"hosted":"my-pub.dev"}'`
+  * Add an sdk dependency:
+    `$topLevelProgram pub add 'foo{"sdk":"flutter"}'`
+  * Add a git dependency:
+    `$topLevelProgram pub add 'foo{"git":"https://github.com/foo/foo"}'`
+  * Add a git dependency with a path and ref specified:
+    `$topLevelProgram pub add 'foo{"git":{"url":"../foo.git","ref":"branch","path":"subdir"}}'`''';
+  @override
+  String get docUrl => 'https://dart.dev/tools/pub/cmd/pub-add';
 
   AddCommand() {
-    argParser.addFlag('dev',
-        abbr: 'd',
-        negatable: false,
-        help: 'Adds to the development dependencies instead.');
-
-    argParser.addOption('git-url', help: 'Git URL of the package');
-    argParser.addOption('git-ref',
-        help: 'Git branch or commit to be retrieved');
-    argParser.addOption('git-path', help: 'Path of git package in repository');
-    argParser.addOption('hosted-url', help: 'URL of package host server');
-    argParser.addOption('path', help: 'Add package from local path');
-    argParser.addOption('sdk',
-        help: 'add package from SDK source',
-        allowed: ['flutter'],
-        valueHelp: '[flutter]');
     argParser.addFlag(
-      'example',
-      help:
-          'Also update dependencies in `example/` after modifying pubspec.yaml in the root package (if it exists).',
+      'dev',
+      abbr: 'd',
+      negatable: false,
+      help: 'Adds to the development dependencies instead.',
+      hide: true,
+    );
+
+    argParser.addOption(
```

Comment:
All of these options need a comment explaining why we're hid them.
```suggestion
    // Following options are hidden/deprecated in favor of the new syntax: [dev:]<package>[:descriptor] ...
    // To avoid breaking changes we keep supporting them, but hide them from --help to discourage
    // further use. Combining these with new syntax will fail.
    argParser.addOption(
```